manager/encryption/encryption.go

```diff
@@ -59,6 +60,51 @@ func (n noopCrypter) Algorithm() api.MaybeEncryptedRecord_Algorithm {
 // decrypt any data
 var NoopCrypter = noopCrypter{}
 
+// specificDecryptor represents a specific type of Decrypter, like NaclSecretbox or Fernet.
+// It does not apply to a more general decrypter like MultiDecrypter.
+type specificDecrypter interface {
+	Decrypter
+	Algorithm() api.MaybeEncryptedRecord_Algorithm
+}
+
+// MultiDecrypter is a decrypter that will attempts to decrypt with multiple decrypters
+type MultiDecrypter map[api.MaybeEncryptedRecord_Algorithm][]Decrypter
```

Comment:
I think newtyping maps and then defining methods on them is... a recipe for confusion later on. I was already confused on line 97 when I saw us iterating over what seemed to me to be an object.

Is there any advantage to newtyping a map over defining a struct with a map field?

Reply:
Also, I am unclear why we need a slice of `Decrypter`s for each algorithm. It seems to me like mapping 1:1 algorithim->decrypter would be sufficient?

Reply:
I agree with @dperny here. Would prefer to have a struct with a map here.

Reply:
> Also, I am unclear why we need a slice of Decrypters for each algorithm. It seems to me like mapping 1:1 algorithim->decrypter would be sufficient?

It's possible to have multiple keys per algorithm, if you have an old and new key.  This is needed for instance when the encryption key is rotated, which happens when you go from unlocked (when the key was stored unencrypted in the TLS key header) to locked (when the key is encrypted in the TLS key header).  Not everything may have finished porting over to use the new key, so some WALs will be encrypted with the old key, and some with the new.  So 1 decrypter with the  new key, 1 with the old, but same algorithm.

Reply:
That makes perfect sense. Could you add a comment to that effect?

Reply:
Done, and updated to a struct